### PR TITLE
Add initial extension implementation.

### PR DESCRIPTION
**Pull request by cevans87**

### CODE REVIEW

File: gh-push [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #!/usr/bin/env bash
 
-trap "cd $(pwd)" EXIT
+HEMLOCK_PRE_PUSH_CLONE_PATH=.pre_push_clone
+_CWD=$(pwd)
```
Comment (jasone) [inline gh-push:4]:
```suggestion
_CWD="$(pwd)"
```
Quoting to support whitespace (here and elsewhere)?

File: gh-push [hunk 2]
```diff
@@ -10,6 +13,14 @@ then
    exit $?
 fi
 
+# Docker pre-push image requires this clone.
+rm -rf ${HEMLOCK_PRE_PUSH_CLONE_PATH}
+git clone . ${HEMLOCK_PRE_PUSH_CLONE_PATH}
+# Remove these files that aren't stable across multiple clone commands to avoid cache busting of the
+# docker build command.
+rm -rf ${HEMLOCK_PRE_PUSH_CLONE_PATH}/.git/logs
+rm ${HEMLOCK_PRE_PUSH_CLONE_PATH}/.git/index
```
Comment (jasone) [inline gh-push]: This seems brittle to me, in that it's not obvious that this is a stable/comprehensive list of things to remove. How about creating a bare repository in the pre-image stage, then cloning a working repo from that as part of the build/test process within the container? Something like the following should capture what we need and omit any ephemeral data which would thwart caching.

```sh
git clone --single-branch --no-tags -b <pr_branch> --bare <src> <dst>
```

Comment (cevans87) [reply gh-push]: I had to make an associated change at https://github.com/BranchTaken/Hemlock/pull/182/files#diff-fab244873dec173ad617d4e26b0eb9814ae799d8abbc8ed8071d2cdbdd5ec0c6R41-R43

**Comment by jasone** [comment]
It looks like you pushed this branch directly to the main repo. Create a fork?

**Review by jasone — APPROVED** [review]

File: gh-push [hunk 3]
```diff
@@ -10,6 +13,11 @@ then
    exit $?
 fi
 
+# Docker pre-push image requires this clone.
+rm -rf ${HEMLOCK_PRE_PUSH_CLONE_PATH}
+git clone --single-branch --no-tags -b $(git rev-parse --abbrev-ref HEAD) --bare \
+    . ${HEMLOCK_PRE_CLONE_PATH}
```
Comment (cevans87) [inline gh-push:19]: This is the main culprit for the broken CI. The env variable isn't defined. It should be `HEMLOCK_PRE_PUSH_CLONE_PATH`.

**Pull request merged**